python/cudf/cudf/core/dtypes.py

```diff
@@ -103,3 +105,46 @@ def deserialize(cls, header, frames):
             categories_header, categories_frames
         )
         return cls(categories=categories, ordered=ordered)
+
+
+class ListDtype(ExtensionDtype):
+    def __init__(self, value_type):
+        if isinstance(value_type, ListDtype):
+            self._typ = pa.list_(value_type._typ)
+        else:
+            value_type = cudf.utils.dtypes.np_to_pa_dtype(np.dtype(value_type))
+            self._typ = pa.list_(value_type)
+
+    @property
+    def value_type(self):
+        if isinstance(self._typ.value_type, pa.ListType):
+            return ListDtype.from_arrow(self._typ.value_type)
+        else:
+            return np.dtype(self._typ.value_type.to_pandas_dtype()).name
+
+    @property
+    def type(self):
+        return list
```

Comment:
I really don't like this because we go from a nice memory representation to a Python list of PyObjects. Any alternative ideas?

Reply:
I'd suggest numpy or cupy arrays here.

Reply:
Ultimately you could imagine a user using this API to create a strongly typed scalar, i.e. `np.dtype('int32').type(5)`

The problem with a CuPy array is then we can't represent strings or datetimes or other types that cuDF supports. Numpy is more flexible but you still can't represent nested types properly.

A pyarrow array is flexible enough to capture this, but we'd really want a pyarrow scalar, which can't be constructed outside of a pyarrow array.

Ultimately I'd be in favor of this returning a `cudf.Scalar` object which is backed by device memory, but I don't think that's feasible now.

Reply:
So leave as list and mark as a TODO?

Reply:
What if we had it raise a `NotImplementedError` for now? Would that break things?

Reply:
> A pyarrow array is flexible enough to capture this, but we'd really want a pyarrow scalar, which can't be constructed outside of a pyarrow array.

This is a good point and I think it's sensible that if the elements of the underlying column follow the arrow memory layout, so should the scalar representation even if it's empty. Couldn't we make it so that if you passed it a scalar or 1-D of length N list you'd get a one-D array with either one or N elements, and if you passed it something nested you'd get the corresponding nested pyarrow array?

Reply:
> Couldn't we make it so that if you passed it a scalar or 1-D of length N list you'd get a one-D array with either one or N elements, and if you passed it something nested you'd get the corresponding nested pyarrow array?

The problem with this is the `.type` property I believe expects a `type` object as the output, and the only way to accomplish this would be to create our own `type` object that on initialization handles creation of the PyArrow array and whatnot. I think this is at least partially similar to `pandas.core.dtypes.dtypes.CategoricalDtypeType` which is what's returned from a `pd.CategoricalDtype(...).type`.

Reply:
Ah, I had hoped the basic pyarrow array constructor would basically cover all of this - I know it can handle instantiation of the correct array from a lot of cases involving nested python lists, but it looks like it doesn't handle scalars gracefully:

```
>>> pa.array(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/array.pxi", line 269, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 38, in pyarrow.lib._sequence_to_array
TypeError: 'int' object is not iterable
```
So you're right we would need to roll something of our own here and I agree we might as well handle that more generally later.

Reply:
> What if we had it raise a NotImplementedError for now? Would that break things?

This does break things. Specifically, we have a few places where we do `hasattr(obj, "type")` (which actually evaluates the property).

Reply:
Would it be really bad to return the result of calling `pa.array(the_arg)` or something similar here, with one extra branch of logic to check `if is_scalar(the_arg)` ?

Reply:
Also shamelessly shilling `cudf.dtype` which would eliminate the need to check if the dtype carried by valid cuDF objects has a `type` property.

Reply:
> Would it be really bad to return the result of calling pa.array(the_arg) or something similar here

Not sure I understand -- what is `the_arg` here?

> Also shamelessly shilling cudf.dtype which would eliminate the need to check if the dtype carried by valid cuDF objects has a type property.

I think if we did decide to roll our own dtype interface, I imagine we'd still want something like a `.type` property that would return something like `cudf.ListScalar` (when we have that).

Reply:
I guess I mean something like this. 
```
class _ListType(object):
    def __init__(self):
        return
    def __call__(self, arg):
        if is_scalar(arg):
           return pa.array([arg])
        else:
            return pa.array(arg)
```
Then we might be able to just do 

```
@property
def type(self):
    return _ListType()
```

Though I could just be unnecessarily complicating things here.

Reply:
Ah - gotcha. We could just return `pa.array` instead of `list` for now. I don't think we should transparently convert scalars to lists though, since neither PyArrow (as demonstrated above) nor Python allow this:

```python
>>> list(1)
TypeError: 'int' object is not iterable
```

Ultimately though, we want whatever we return to match the type of  `ListColumn[index]`, which will eventually be a `cudf.Scalar`.

Reply:
@brandon-b-miller @kkraus14 any preference between `list` and `pa.array` for now (until we support `cudf.Scalar` more widely)?

Reply:
I think the downside of `list` is that it maybe implies that sequences of general python objects are valid within the cuDF. The arrow array constructor seems a little more consistent with what's going on under the hood in my opinion.

Reply:
👍  thanks Brandon - made it a `pa.array`